L82-core-event-engine.md

```diff
@@ -0,0 +1,478 @@
+
+There are three ways in which EventEngines can be provided to gRPC.
+
+#### Setting an EventEngine instance for a Channel
+
+EventEngines can be configured per channel, which allows customization for both
+clients and servers. To set a custom EventEngine for a client channel, you can
+do something like the following:
+
+```cpp
+ChannelArguments args;
+std::shared_ptr<EventEngine> engine = std::make_shared<MyEngine>(...);
+args.SetEventEngine(engine);
+MyAppClient client(grpc::CreateCustomChannel(
+    "localhost:50051", grpc::InsecureChannelCredentials(), args));
+```
+
+#### Setting an EventEngine instance for a Server
+
+A gRPC server can use a custom EventEngine by calling the
+`ServerBuilder::SetEventEngine` method:
+
+```cpp
+ServerBuilder builder;
+std::shared_ptr<EventEngine> engine = std::make_shared<MyEngine>(...);
+builder.SetEventEngine(engine);
+std::unique_ptr<Server> server(builder.BuildAndStart());
+server->Wait();
+```
+
+#### Providing a default EventEngine factory
+
+TODO: this has not been thoroughly discussed.
+
+Before gRPC is initialized, an application can specify an EventEngine Factory
+method that will be used within gRPC when a channel- or server-specific
+EventEngine is unavailable. gRPC needs to perform I/O, run tasks, and use timers
+in scenarios that are not directly attached to channels and servers.  Providing
+a custom EventEngine factory will allow your application to avoid the creation
+of the default EventEngine provided by gRPC.
+
+### Lifetime and Ownership
+
+gRPC will take shared ownership of application-provided EventEngines via
+std::shared\_ptrs to ensure that the engines remain available until they are no
+longer needed. Depending on the use case, EventEngines may need to live until
+gRPC is shut down. Any EventEngine that is provided to gRPC is expected to
+remain valid for as long as gRPC holds a pointer to it. Application-provided
+EventEngines will not be shut down automatically when gRPC is shut down, it is
+the application's responsibility to shut down and clean up EventEngines when
+they are no longer needed.
+
+### Temporary protections
+
+The EventEngine can be enabled at compile time by defining the C/C++ macro
+`GRPC_USE_EVENT_ENGINE`. EventEngines will not be enabled by default on any
+platform to begin with.
+
+The EventEngine API will exist in the `grpc_event_engine::experimental`
+namespace until it has stabilized. Stabilization will likely be signaled by
+having all platforms supported sufficiently by the default EventEngine, and
+having all other primary EventEngine use cases exercised without needing to make
+further API changes.
+
+The EventEngine rollout is planned to be on a per-platform basis. Once we
+establish that the default EventEngine implementation is sufficient for a
+specific platform, we will change the build to replace iomgr with EventEngine on
+that platform.
+
+## Limitations
+
+### Wrapped languages cannot provide a custom EventEngine at runtime
```

Comment:
Can we have a discussion on why?

@nathanielmanistaatgoogle made a solid case for allowing such integrations years ago, and I'd hoped that EventEngine could fill the gap in allowing gRPC IO to be implemented by non-C/C++ languages.

Reply:
I hadn't seen any input from @nathanielmanistaatgoogle when researching this project. I'd be interested to hear it!

The answers to "why not" are: early hypotheses about poor performance given the way many wrapped languages are implemented, low ROI (relatively few reports of people wanting to implement custom I/O in wrapped languages), and our overall priorities for the work - we make compromises to minimize complexity and increase the likelihood that this project is shipped where other similar efforts have not.

I'm happy to have the discussion again, I think @markdroth and @nicolasnoble should be involved as they have the most context on previous work and current needs.

Reply:
The main limitations here were mainly dll-barrier crossing wrt to memory management, and, well, C++.

Reply:
The proposals were around sans-io and allowing more event loops to leverage gRPC in a performant manner.
I'm not sure any of the stated objections would actually bear out preventing an implementation, and so my preference here would be to simply omit this section - maybe with a note that we don't plan on exposing any of this to wrapped languages other than C++ at this point in time.

Reply:
Python has many heterogeneous asynchronous library implementation, they use different IO libraries. There were asks from time to time. Based on the event-engine-setting API above, what is blocking wrapper languages to provide custom EventEngine?

Nico mentioned memory management issue, then if several EventEngines are provided by Python, users still have the chance to select at the runtime right?

If above works, the wrapper language can create a shell EventEngine to invoke user-provided logic, then custom EventEngine is doable?

Reply:
I've weakened the statement here to state that wrapped language support for writing and providing custom EventEngines is not an immediate goal of this work. I'd like to leave this section in, since the scope of wrapped language support is an obvious question that I figure folks would want an answer to.

@lidizheng I think there are mostly just a broader set of concerns to address if we want to move EventEngine support up closer to the wrapped language layers. gRPC will allow implementations of the C++ EventEngine API to be plugged in at runtime; if the Python layer has support for multiple EventEngines written in C/C++ (I assume compiled in and distributed with our python packages), you could write a Python API to apply them as needed. To your second question on whether a proxying C++ EventEngine could work by ushering C-core requests to/from EventEngine-like things written in Python: probably yes.